pkg/build/strategies/onbuild/onbuild.go

```diff
@@ -87,28 +95,33 @@ func (builder *OnBuild) SourceTar(config *api.Config) (io.ReadCloser, error) {
 
 // Build executes the ONBUILD kind of build
 func (builder *OnBuild) Build(config *api.Config) (*api.Result, error) {
+	buildResult := &api.Result{}
+
 	if config.BlockOnBuild {
-		return nil, fmt.Errorf("builder image uses ONBUILD instructions but ONBUILD is not allowed")
+		buildResult.BuildInfo = updater.UpdateFailureReason(buildResult.BuildInfo, api.ReasonGenericS2IBuildFailed)
```

Comment:
In line with my other comment, but perhaps counter to the "you do not need to be completely exhaustive" guidance @bparees gave, but my initial reaction is that if we don't have a specific error, at least a generic "config mismatch" reason would be good.